Python/traceback.c

```diff
@@ -554,29 +697,53 @@ tb_displayline(PyTracebackObject* tb, PyObject *f, PyObject *filename, int linen
         if (start_col_byte_offset < 0 || end_col_byte_offset < 0) {
             goto done;
         }
+
         // Convert the utf-8 byte offset to the actual character offset so we
         // print the right number of carets.
-        Py_ssize_t start_offset = _PyPegen_byte_offset_to_character_offset(source_line, start_col_byte_offset);
-        Py_ssize_t end_offset = _PyPegen_byte_offset_to_character_offset(source_line, end_col_byte_offset);
+        Py_ssize_t start_offset = (Py_ssize_t)start_col_byte_offset;
+        Py_ssize_t end_offset = (Py_ssize_t)end_col_byte_offset;
 
-        char offset = truncation;
-        while (++offset <= start_offset) {
-            err = PyFile_WriteString(" ", f);
-            if (err < 0) {
-                goto done;
+        if (source_line) {
+            start_offset = _PyPegen_byte_offset_to_character_offset(source_line, start_col_byte_offset);
+            end_offset = _PyPegen_byte_offset_to_character_offset(source_line, end_col_byte_offset);
+        }
+
+        const char *primary, *secondary;
+        primary = secondary = "^";
+
+        int left_end_offset, right_start_offset;
+        left_end_offset = right_start_offset = Py_SAFE_DOWNCAST(end_offset, Py_ssize_t, int) - Py_SAFE_DOWNCAST(start_offset, Py_ssize_t, int);
+
+        if (source_line) {
+            int res = extract_anchors_from_line(filename, source_line, start_offset, end_offset,
+                                                &left_end_offset, &right_start_offset);
+            if (res < 0) {
+                err = ignore_source_errors();
+                if (err < 0) {
+                    goto done;
+                }
+            } else if (res > 0) {
+                primary = "^";
+                secondary = "~";
             }
         }
-        while (++offset <= end_offset + 1) {
-            err = PyFile_WriteString("^", f);
-            if (err < 0) {
-                goto done;
+
+        char offset = truncation;
+        while (++offset <= end_offset) {
+            if (offset <= start_offset) {
+                err = PyFile_WriteString(" ", f);
+            } else if (offset <= left_end_offset + start_offset) {
+                err = PyFile_WriteString(secondary, f);
+            } else if (offset <= right_start_offset + start_offset) {
+                err = PyFile_WriteString(primary, f);
+            } else {
+                err = PyFile_WriteString(secondary, f);
             }
         }
         err = PyFile_WriteString("\n", f);
     }
-
     else {
```

Comment:
Hmm, since we're adding more and more code into this if block, I wonder if we should make the error condition an early return to organize this better and make it less indented.

```c
    /* ignore errors since we can't report them, can we? */
    if (_Py_DisplaySourceLine(f, filename, lineno, _TRACEBACK_SOURCE_LINE_INDENT,
                              &truncation, &source_line)) {
        err = ignore_source_errors();
        goto done;
    }
    
    int code_offset = tb->tb_lasti;
    PyCodeObject* code = _PyFrame_GetCode(frame);

    int start_line;
    int end_line;
    int start_col_byte_offset;
    int end_col_byte_offset;
    ...
```

Reply:
Slowly getting there ;-) (from a possible upcoming patch)
https://github.com/isidentical/cpython/blob/5eb3a24a87328ca9f1ff03aa8cc206c731553b94/Python/traceback.c#L570-L584